src/shared/tools.cpp

```diff
@@ -4,36 +4,31 @@
 
 ////////////////////////// rnd numbers ////////////////////////////////////////
 
-#define N (624)
-#define M (397)
-#define K (0x9908B0DFU)
+std::random_device rndseed;
+std::ranlux48 rndalg(rndseed());
 
-static uint state[N];
-static int next = N;
+int rnd(int value)
+{
+    std::uniform_int_distribution<> rndint(0, value-1);
+    return rndint(rndalg);
+}
 
-void seedMT(uint seed)
+float rndfloat(int value)
 {
-    state[0] = seed;
-    for(uint i = 1; i < N; i++)
-        state[i] = seed = 1812433253U * (seed ^ (seed >> 30)) + i;
-    next = 0;
+    std::uniform_real_distribution<> rndreal(0, value-1);
+    return rndreal(rndalg);
 }
 
-uint randomMT()
+int detrnd(int seed, int value)
 {
-    int cur = next;
-    if(++next >= N)
-    {
-        if(next > N) { seedMT(5489U + time(NULL)); cur = next++; }
-        else next = 0;
-    }
-    uint y = (state[cur] & 0x80000000U) | (state[next] & 0x7FFFFFFFU);
-    state[cur] = y = state[cur < N-M ? cur + M : cur + M-N] ^ (y >> 1) ^ (-int(y & 1U) & K);
-    y ^= (y >> 11);
-    y ^= (y <<  7) & 0x9D2C5680U;
-    y ^= (y << 15) & 0xEFC60000U;
-    y ^= (y >> 18);
-    return y;
+    std::ranlux48 algseed(seed);
```

Comment:
I understand `detrnd` is supposed to mean "deterministic random number", but I'm not sure what properties the implementation is supposed to provide. We might want to check its uses. The only reason I could come up with where you'd want to specify your own seed is to get the same results on more than one host. But that seems like a bad idea.

Regarding the reproducibility of the results of `std::uniform_int_distribution`, it is not stable over different implementations. While it will always return the same sequence of numbers in, e.g., libstdc++, libc++ differs.

Here's an example:

```cpp
#include <iostream>
#include <random>

int main() {
    // https://xkcd.com/221/
    std::mt19937 rnd(4);
    std::uniform_int_distribution<> dist(0, (1ULL << 16));
    std::cout << dist(rnd) << std::endl;
}
```

With GCC 7.5.0 with libstdc++, it prints 63376. With Clang 9 and libc++, it gives you 11863. So, the deterministic property of this little function is only provided if you can guarantee the same STL implementation is used. Hence, depending on the purpose of this function, the entire concept should be reviewed in the context of Blue Nebula.

Might be worth creating a new issue.

Reply:
afaik `detrnd` provides more "consistent" and thus "closer" results than `rnd`, which for example results useful for picking a certain range of colors for exploding mines (which tend to turn red when using `rnd`) or the flashes on the player which holds the rocket (it's way too fast with `rnd`), so it's not about getting the same results.
`std::uniform_int_distribution` is a non-issue as the results don't need to be reproducible between different STL implementations

Reply:
> afaik detrnd provides more "consistent" and thus "closer" results than rnd

Well, it has to do something with predictability, obviously, as the seed is defined by the using code. The intention must be to get reproducibe results, at least on the same machine.

If you feed `detrnd` with the same seed (which is, e.g., calculated from position data, in some function), then you'll get the same result, at least on that machine. I don't understand why that's needed, though. There must be reason for it.

Can you point me to where these colors are calculated, please?

Reply:
Player pulsing (also affects burn and shock):
https://github.com/blue-nebula/base/blob/0da5037bbdd63df8c377cddc1e2c4675f5353e16/src/game/game.cpp#L842

Mines and nades:
https://github.com/blue-nebula/base/blob/0da5037bbdd63df8c377cddc1e2c4675f5353e16/src/engine/lightning.h#L41